package World;

message World {

    message DoubleRow {
        repeated double cells = 1;
    }

    message BooleanRow {
        repeated bool cells = 1;
    }

    message IntegerRow {
        repeated int32 cells = 1;
    }

    message ByteRow {
        repeated int32 cells = 1;
    }

    message DoubleMatrix {
        repeated DoubleRow rows = 1;
    }

    message BooleanMatrix {
        repeated BooleanRow rows = 1;
    }    

    message IntegerMatrix {
        repeated IntegerRow rows = 1;
    }   

    message DoubleQuantile {
        required int32 key = 1;
        required double value = 2;
    }

    message DoubleMatrixWithQuantiles {
        repeated DoubleQuantile quantiles = 1;
        repeated DoubleRow rows = 2;
    }     

    required string name   = 1;
    required int32  width  = 2;
    required int32  height = 3;

    // Elevation

    required DoubleMatrix heightMapData  = 4;
    required double heightMapTh_sea      = 5;
    required double heightMapTh_plain    = 6;
    required double heightMapTh_hill     = 7;

    // Ocean

    required BooleanMatrix ocean         = 8;

    // Biome
    optional IntegerMatrix biome         = 9;

    // Humidity
    optional DoubleMatrixWithQuantiles humidity = 10;

    // Irrigation
    optional DoubleMatrix irrigation  = 11;

    // Permeability
    optional DoubleMatrix permeabilityData = 12;
    optional double permeability_low       = 13;
    optional double permeability_med       = 14;

    // Watermap
    optional DoubleMatrix watermapData     = 15;
    optional double watermap_creek         = 16;
    optional double watermap_river         = 17;
    optional double watermap_mainriver     = 18;

    // Precipitation
    optional DoubleMatrix precipitationData = 19;
    optional double precipitation_low       = 20;
    optional double precipitation_med       = 21;    

    // Temperature
}


